From: Hanne-Lotta Mäenpää <hannelotta@gmail.com>
Subject: headers: add check for C standard version
Date: Sun, 26 Oct 2025 21:58:46 +0200

Compiling the kernel with GCC 15 results in errors, as with GCC 15 the
default language version for C compilation has been changed from
-std=gnu17 to -std=gnu23 - unless the language version has been changed
using

    KBUILD_CFLAGS += -std=gnu17

or earlier.

C23 includes new keywords 'bool', 'true' and 'false', which cause
compilation errors in Linux headers:

    ./include/linux/types.h:30:33: error: `bool' cannot be defined
        via `typedef'

    ./include/linux/stddef.h:11:9: error: cannot use keyword `false'
        as enumeration constant

Add check for C Standard's version in the header files to be able to
compile the kernel with C23.

Link: https://lkml.kernel.org/r/20251026195846.69740-1-hannelotta@gmail.com
Signed-off-by: Hanne-Lotta Mäenpää <hannelotta@gmail.com>
Cc: David Hunter <david.hunter.linux@gmail.com>
Cc: "Gustavo A. R. Silva" <gustavoars@kernel.org>
Cc: Kees Cook <kees@kernel.org>
Cc: Arnd Bergmann <arnd@arndb.de>
Cc: <stable@vger.kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/stddef.h |    2 ++
 include/linux/types.h  |    2 ++
 2 files changed, 4 insertions(+)

--- a/include/linux/stddef.h~headers-add-check-for-c-standard-version
+++ a/include/linux/stddef.h
@@ -7,10 +7,12 @@
 #undef NULL
 #define NULL ((void *)0)
 
+#if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 202311L
 enum {
 	false	= 0,
 	true	= 1
 };
+#endif
 
 #undef offsetof
 #define offsetof(TYPE, MEMBER)	__builtin_offsetof(TYPE, MEMBER)
--- a/include/linux/types.h~headers-add-check-for-c-standard-version
+++ a/include/linux/types.h
@@ -32,7 +32,9 @@ typedef __kernel_timer_t	timer_t;
 typedef __kernel_clockid_t	clockid_t;
 typedef __kernel_mqd_t		mqd_t;
 
+#if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 202311L
 typedef _Bool			bool;
+#endif
 
 typedef __kernel_uid32_t	uid_t;
 typedef __kernel_gid32_t	gid_t;
_
